src/providers/pkcs11/mod.rs

```diff
@@ -74,10 +74,23 @@ impl Provider {
     fn new(
         key_info_store: KeyInfoManagerClient,
         backend: Pkcs11,
-        slot_number: Slot,
+        slot_number: Option<Slot>,
         user_pin: Option<SecretString>,
         software_public_operations: bool,
     ) -> Option<Provider> {
+        // If slot_number is not valid take the first available with a valid token
+        let slot_number = match slot_number {
+            Some(i) => i,
+            None => {
+                let slots = backend.get_slots_with_token().ok()?;
+                if !slots.is_empty() {
+                    slots[0]
```

Comment:
Actually, I'm wondering - should we do it this way, grabbing "the first" token, or should we fail if `slots.len() != 1`? My worry is whether the result is deterministic - what if there are two slots and you get them arranged at random? Have to admit, I've not looked at the spec too closely.

Reply:
(or, what if a new slot/token is added at some later point and that one becomes the first one? your service will spin up, but you'll essentially lose a lot of keys)

Reply:
Yes that's a very good point and I think a good suggestion. I checked the PKCS 11 docs and nothing says that the order is deterministic but also because of the other reason you indicated I think it's better to check that there is only one slot to use.

So replacing the anove with:
```diff
-if !slots.is_empty() {
+if slots.len() == 1 {
```

We should make it clear that this behaviour is used in the config file comments.